### PR TITLE
Fix contract violation in debugger func eval

### LINKED ISSUE #54504: Contract Violation when attaching debugger in VS
I encountered this CoreCLR contract violation when I attached the VS debugger to a debug CoreCLR build with a Just-In-Time Debugging session:

```
Assert failure(PID 12088 [0x00002f38], Thread: 25700 [0x6464]): CONTRACT VIOLATION by Debugger::GetInteropSafeExecutableHeap at "D:\source\runtime\src\coreclr\debug\ee\debugger.cpp" @ 1529

THROWS called in a NOTHROW region.

                        CONTRACT in Debugger::GetInteropSafeExecutableHeap at "D:\source\runtime\src\coreclr\debug\ee\debugger.cpp" @ 1529VIOLATED-->  CONTRACT in Debugger::FuncEvalSetup at "D:\source\runtime\src\coreclr\debug\ee\debugger.cpp" @ 15082
                        CONTRACT in Debugger::HandleIPCEvent at "D:\source\runtime\src\coreclr\debug\ee\debugger.cpp" @ 10501
                        CONTRACT in DebuggerRCThread::HandleRSEA at "D:\source\runtime\src\coreclr\debug\ee\rcthread.cpp" @ 838
                        CONTRACT in DebuggerRCThread::MainLoop at "D:\source\runtime\src\coreclr\debug\ee\rcthread.cpp" @ 908
                        CONTRACT in DebuggerRCThread::ThreadProc at "D:\source\runtime\src\coreclr\debug\ee\rcthread.cpp" @ 652



CORECLR! CONTRACT_ASSERT + 0x3C8 (0x00007fff`f56f9098)
CORECLR! ClrDebugState::CheckOkayToThrow + 0x55 (0x00007fff`f56fe645)
CORECLR! BaseContract::DoChecks + 0x2BE (0x00007fff`f570833e)
CORECLR! EEContract::DoChecks + 0x8A (0x00007fff`f5c63e9a)
CORECLR! Debugger::GetInteropSafeExecutableHeap + 0x13B (0x00007fff`f6298efb)
CORECLR! DebuggerEval::DebuggerEval + 0x6A (0x00007fff`f6284e6a)
CORECLR! Debugger::FuncEvalSetup + 0x3E6 (0x00007fff`f6294596)
CORECLR! Debugger::HandleIPCEvent + 0x1E30 (0x00007fff`f629f720)
CORECLR! `HandleIPCEventWrapper'::`3'::__Body::Run + 0x65 (0x00007fff`f62e5d65)
CORECLR! `HandleIPCEventWrapper'::`3'::__Body::Run + 0x65 (0x00007fff`f62e5d65)
CORECLR! HandleIPCEventWrapper + 0x75 (0x00007fff`f62e37d5)
    File: D:\source\runtime\src\coreclr\debug\ee\debugger.cpp Line: 1529
    Image: D:\source\runtime\artifacts\tests\coreclr\windows.x64.Debug\Tests\Core_Root\corerun.exe
```

### PR DESCRIPTION
**Pull request by davidwrighton**
`DebuggerEval::DebuggerEval` is called from `Debugger::FuncEvalSetup` which is a NOTHROW function. As the `GetInteropSafeExecutableHeap` and associated `Alloc` functions can fail with an OOM, move the allocation out of the constructor to the caller where we can safely return `E_OUTOFMEMORY`.

This contract violation prevents debugging on chk and debug builds of the runtime.

Fixes #54504

### CODE REVIEW

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Comment by jkoritzinsky** [comment]
@davidwrighton I believe this fixes https://github.com/dotnet/runtime/issues/54504

Is that correct?

**Comment by AaronRobinsonMSFT** [comment]
>Is that correct?

Yes.

**Review by jkoritzinsky — APPROVED** [review]

**Comment by davidwrighton** [comment]
@hoyosjs I'd like a diagnostics dev to sign off on this.

**Comment by hoyosjs** [comment]
Looking at this. Checking the RW path

File: src/coreclr/debug/ee/debugger.cpp [hunk 1]
```diff
@@ -1315,13 +1315,10 @@ ULONG DebuggerMethodInfoTable::CheckDmiTable(void)
...
+DebuggerEval::DebuggerEval(CONTEXT * pContext, DebuggerIPCE_FuncEvalInfo * pEvalInfo, bool fInException, DebuggerEvalBreakpointInfoSegment* bpInfoSegmentRX)
 {
     WRAPPER_NO_CONTRACT;
 
-    // Allocate the breakpoint instruction info in executable memory.
-    void *bpInfoSegmentRX = g_pDebugger->GetInteropSafeExecutableHeap()->Alloc(sizeof(DebuggerEvalBreakpointInfoSegment));
-
 #if !defined(DBI_COMPILE) && !defined(DACCESS_COMPILE) && defined(HOST_OSX) && defined(HOST_ARM64)
     ExecutableWriterHolder<DebuggerEvalBreakpointInfoSegment> bpInfoSegmentWriterHolder((DebuggerEvalBreakpointInfoSegment*)bpInfoSegmentRX, sizeof(DebuggerEvalBreakpointInfoSegment));
     DebuggerEvalBreakpointInfoSegment *bpInfoSegmentRW = bpInfoSegmentWriterHolder.GetRW();
```
Comment (hoyosjs) [inline src/coreclr/debug/ee/debugger.cpp:1324]: Not a big thing, minor nit: can you remove the castings on 1323, 1326, and 1329 now that it's the right type of parameter?

File: src/coreclr/debug/ee/debugger.cpp [hunk 2]
```diff
@@ -1315,13 +1315,10 @@ ULONG DebuggerMethodInfoTable::CheckDmiTable(void)
 //      pEvalInfo - Contains all the important information, such as parameters, type args, method.
 //      fInException - TRUE if the thread for the eval is currently in an exception notification.
 //
-DebuggerEval::DebuggerEval(CONTEXT * pContext, DebuggerIPCE_FuncEvalInfo * pEvalInfo, bool fInException)
+DebuggerEval::DebuggerEval(CONTEXT * pContext, DebuggerIPCE_FuncEvalInfo * pEvalInfo, bool fInException, DebuggerEvalBreakpointInfoSegment* bpInfoSegmentRX)
```
Comment (hoyosjs) [inline src/coreclr/debug/ee/debugger.cpp:1318]: Can you please add a comment for the new parameter that the `bpInfoSegmentRX` is a InteropSafe allocation allocated by caller but with lifetime handled by this class please?

**Review by hoyosjs — APPROVED** [review]

**Pull request merged**